Add snowbridge transformation testing guide (#401)
snowplow/documentation: docs/destinations/forwarding-events/snowbridge/configuration/transformations/testing/index.md

---
title: "Testing Transformations"
date: "2023-04-14"
sidebar_position: 300
---

```mdx-code-block
import {versions} from '@site/src/componentVersions';
import CodeBlock from '@theme/CodeBlock';
```

> The easiest way to test a transformation configuration for Snowbridge is to run it locally, using the default stdin source, stdout target and your actual transformation config - using a sample of data that is as close to the real world as possible.

Reviewer: ```suggestion
The easiest way to test a transformation configuration for Snowbridge is to run it locally, using the default `stdin` source, `stdout` target and your actual transformation config - using a sample of data that is as close to the real world as possible.
```

The sample file should contain the events/messages you'd like to test with, one per line.

## Snowplow data - generating a sample

If you're working with Snowplow data, you can follow the [guide to use Snowplow Micro](/docs/getting-started-with-micro/basic-usage/index.md) to generate test data, using the `--output-tsv` to get the data into a file, as per the [exporting to tsv section](/docs/getting-started-with-micro/basic-usage/index.md#exporting-events-to-tsv).

For example, here we're using a file named `data.tsv`:

<CodeBlock language="bash">{
`docker run -p 9090:9090 snowplow/snowplow-micro:${versions.snowplowMicro} --output-tsv > data.tsv`
}</CodeBlock>

Point some test environment tracking to `localhost:9090`, and your events should land in `data.tsv`.

## Testing Snowbridge locally

### Local Snowbridge setup

You can run Snowbridge locally via Docker:

<CodeBlock language="bash">{
`docker run snowplow/snowbridge:${versions.snowbridge}`
}</CodeBlock>

> The default configuration for Snowbridge already uses the `stdin` source and the `stdout` target. So, to test sending data through with no transformations, we can run the following command (where `data.tsv` is a file with Snowplow events in TSV format):T

Reviewer: ```suggestion
The default configuration for Snowbridge already uses the `stdin` source and the `stdout` target. So, to test sending data through with no transformations, we can run the following command (where `data.tsv` is a file with Snowplow events in TSV format):
```

<CodeBlock language="bash">{
`cat data.tsv | docker run -i snowplow/snowbridge:${versions.snowbridge}`
}</CodeBlock>

> This will print the data to console, along with logs.

Reviewer: ```suggestion
This will print the data to the terminal, along with logs.
```
To avoid confusion with BDP Console (who knows)

---
title: "Testing Transformations"
date: "2023-04-14"
sidebar_position: 300
---

```mdx-code-block
import {versions} from '@site/src/componentVersions';
import CodeBlock from '@theme/CodeBlock';
```

The easiest way to test a transformation configuration for Snowbridge is to run it locally, using the default stdin source, stdout target and your actual transformation config - using a sample of data that is as close to the real world as possible.

The sample file should contain the events/messages you'd like to test with, line separated.

## Snowplow data - generating a sample

> If you're working with Snowplow data, you can simpy follow the [guide to use Snowplow Micro](/docs/getting-started-with-micro/basic-usage/index.md) to generate test data, using the `--output-tsv` to get the data into a file, as per the [exporting to tsv section](/docs/getting-started-with-micro/basic-usage/index.md#exporting-events-to-tsv).

Reviewer: ```suggestion
If you're working with Snowplow data, you can follow the [guide to use Snowplow Micro](/docs/getting-started-with-micro/basic-usage/index.md) to generate test data, using the `--output-tsv` to get the data into a file, as per the [exporting to tsv section](/docs/getting-started-with-micro/basic-usage/index.md#exporting-events-to-tsv).
```

For example, here we're using a file named `data.tsv`:

<CodeBlock language="bash">{
`docker run -p 9090:9090 snowplow/snowplow-micro:${versions.snowplowMicro} --output-tsv > data.tsv`
}</CodeBlock>

Point some test environment tracking to `localhost:9090`, and your events should land in `data.tsv`.

## Testing Snowbridge locally

### Local Snowbridge setup

Snowbridge can be run locally via docker, with:

<CodeBlock language="bash">{
`docker run snowplow/snowbridge:${versions.snowbridge}`
}</CodeBlock>

> Since the default configuration for snowbridge is already `stdin` source and `stdout` target. So, to test sending data through with no transformations, (where `data.tsv` contains our test data) we can simply run:

Reviewer: ```suggestion
The default configuration for Snowbridge is already using the `stdin` source and the `stdout` target. So, to test sending data through with no transformations, we can run the following command (where `data.tsv` is a file with Snowplow events in TSV format):
```

<CodeBlock language="bash">{
`cat data.tsv | docker run -i snowplow/snowbridge:${versions.snowbridge}`
}</CodeBlock>

This will print the data to console, along with logs - note that the metrics reported in logs may report that no data has been processed. This is because the app reached the end of output and exited before the default reporting period of 1 second.

> It is likely easiest to output the results to file, to make it easier to examine the results (note that here I use):

Reviewer: ```suggestion
You can output the results to a file to make it easier to examine them:
```

<CodeBlock language="bash">{
`cat data.tsv | docker run -i snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

> Note that the resulting data will contain more than the data: `PartitionKey`, `TimeCreated` `TimePulled` and `TimeTransformed` will also be present. The data that reaches the target in a production setup is under `Data`.

Reviewer: ```suggestion
:::tip

The output (in `output.txt`) will contain more than the data itself. There will be additional fields called `PartitionKey`, `TimeCreated` `TimePulled` and `TimeTransformed`. The data that reaches the target in a production setup is under `Data`.

:::
```

Reviewer: Adding some color to the page

### Adding configuration

> Adding the specific configuration to test is a matter of mounting a configuration file, and setting the `SNOWBRIDGE_CONFIG_FILE` env var to point to that path. If your configuration is in `config.hcl`, you can add configuration to the above example as follows:

Reviewer: ```suggestion
To add a specific configuration to test, create a configuration file (`config.hcl`) and pass it to the Docker container. You will need to mount the file and set the `SNOWBRIDGE_CONFIG_FILE` environment variable to the path _inside_ the container:
```

<CodeBlock language="bash">{

> `cat data.tsv | docker run -i --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl snowplow/snowbridge:${versions.snowbridge} > output.txt`

Reviewer: ```suggestion
`cat data.tsv | docker run -i \\
  --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl \\
  --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl \\
  snowplow/snowbridge:${versions.snowbridge} > output.txt`
```

Reviewer: This is quite arcane but oh well...

Author: wdym? Happy to change it if there's a better way

Reviewer: I mean the way the Docker image itself works — would be easier if it read a config from a fixed dir, like `/config/config.hcl`. Then the user only needs to mount, but not to pass the env var.

Author: Yeah you're right. Made an issue for it: https://github.com/snowplow/snowbridge/issues/271

}</CodeBlock>

> Note that docker expects absolute paths for mounted files - here we use pwd but you can specify the aboslute path manually too.

Reviewer: I would link to the Docker docs on bind mounts — you can copy the approach from the Micro guides

> Note that docker expects absolute paths for mounted files - here we use pwd but you can specify the aboslute path manually too.

Reviewer: ```suggestion
Note that docker expects absolute paths for mounted files - here we use `$(pwd)` but you can specify the aboslute path manually too.
```

It is likely easiest to output the results to file, to make it easier to examine the results:

<CodeBlock language="bash">{
`cat data.tsv | docker run -i snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

Note that the resulting data will contain more than the data: `PartitionKey`, `TimeCreated` `TimePulled` and `TimeTransformed` will also be present. The data that reaches the target in a production setup is under `Data`.

### Adding configuration

Adding the specific configuration to test is a matter of mounting a configuration file, and setting the `SNOWBRIDGE_CONFIG_FILE` env var to point to that path. If your configuration is in `config.hcl`, you can add configuration to the above example as follows:

<CodeBlock language="bash">{
`cat data.tsv | docker run -i \\
    --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl \\
    --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl \\
    snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

Note that docker expects absolute paths for mounted files - here we use pwd but you can specify the aboslute path manually too.

> To test transformations, your configuration file only needs to contain the `transform` block(s) that you're testing, leaving the source and targets on default (by not specifying them in the configuration file).

Reviewer: ```suggestion
To test transformations, you only need to add the `transform` block(s) to your configuration file. Don’t specify the `source` and `target` blocks to leave them on default (`stdin` and `stdout`).
```

### Adding custom scripts

Finally, you can add custom scripts by mounting a file, in the exact same manner. Assuming the script is in `script.js`, that looks like this:

<CodeBlock language="bash">{
`cat data.tsv | docker run -i \\
    --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl \\
    --mount type=bind,source=$(pwd)/script.js,target=/tmp/script.js \\
    --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl \\
    snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

> Note that you must ensure that the transformation config points to the mounted path (the `target` in the mount path - `/tmp/script.js` above).

Reviewer: I would show an excerpt of the config file to make this easier to follow.

> ## Further testing

Reviewer: Should this be “Testing sources and targets”? (Not sure, but thought I would ask)

Author: As written I think that's what it should be named - but actually instead I should change the phrasing, as it's not limited to sources and targets. :)

> You can use the above method to test sources and targets from a local environment too, which can be useful in combination with transformations - simply set up the appropriate authentication for your source and/or target, add them to your configuration file, and run as directed above.

Reviewer: ```suggestion
You can use the above method to test sources and targets from a local environment too. This can be useful in combination with transformations. Set up the appropriate authentication for your source and/or target, add them to your configuration file, and run the commands from the previous sections.
```

---
title: "Testing Transformations"
date: "2023-04-14"
sidebar_position: 300
---

```mdx-code-block
import {versions} from '@site/src/componentVersions';
import CodeBlock from '@theme/CodeBlock';
```

The easiest way to test a transformation configuration for Snowbridge is to run it locally, using the default stdin source, stdout target and your actual transformation config - using a sample of data that is as close to the real world as possible.

> The sample file should contain the events/messages you'd like to test with, line separated.

Reviewer: ```suggestion
The sample file should contain the events/messages you'd like to test with, one per line.
```

<CodeBlock language="bash">{
`cat data.tsv | docker run -i \\
    --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl \\
    --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl \\
    snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

Note that docker expects absolute paths for mounted files - here we use `$(pwd)` but you can specify the aboslute path manually too.

To test transformations, you only need to add the `transform` block(s) to your configuration file. Don’t specify the `source` and `target` blocks to leave them on default (`stdin` and `stdout`).

### Adding custom scripts

Finally, you can add custom scripts by mounting a file, in the exact same manner. Assuming the script is in `script.js`, that looks like this:

<CodeBlock language="bash">{
`cat data.tsv | docker run -i \\
    --mount type=bind,source=$(pwd)/config.hcl,target=/tmp/config.hcl \\
    --mount type=bind,source=$(pwd)/script.js,target=/tmp/script.js \\
    --env SNOWBRIDGE_CONFIG_FILE=/tmp/config.hcl \\
    snowplow/snowbridge:${versions.snowbridge} > output.txt`
}</CodeBlock>

> Note that you must ensure that the transformation config points to the mounted path (the `target` in the mount path - `/tmp/script.js` above). For example, the transformation block in our cofiguration for the above example might look like this:

Reviewer: ```suggestion
The transformation config should point to the path of the script _inside_ the container (`/tmp/script.js` above). For example, the transformation block in the configuration might look like this:
```

```hcl
transform {
  use "js" {

>     script_path = env.JS_SCRIPT_PATH

Reviewer: Where does `JS_SCRIPT_PATH` come from?

Author: 🤦 meant to change, am dumb, will fix :)

  }
}
```

## Further testing

> You can use the above method to test all other aspects of the app from a local environment too, including sources, targets, failure targets, metrics endpoints etc. To do so, first you'll need to ensure that the local envionment has access to any required resources and can authenticate (eg. connecting from a laptop to a cloud account, or setting up a local metrics server for testing). Once that's done, provide Snowbridge with a hcl file configuring it to connect to those resources, and run it the same way we do in the transformation examples above.

Reviewer: ```suggestion
You can use the above method to test all other aspects of the app from a local environment too, including sources, targets, failure targets, metrics endpoints etc. To do so, first you'll need to ensure that the local envionment has access to any required resources and can authenticate (eg. connecting from a laptop to a cloud account, or setting up a local metrics server for testing). Once that's done, provide Snowbridge with an hcl file configuring it to connect to those resources, and run it the same way as in the transformation examples above.
```
